daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1104,10 +1106,21 @@ private[lf] object SBuiltin {
       val (actualTemplateId, record) = getSAnyContract(args, 1)
 
       if ((templateId +: acceptedTemplateIds).contains(actualTemplateId)) {
-        // TODO: https://github.com/digital-asset/daml/issues/16151
-        // Later, this will need to extend values of predecessor template
-        // types (e.g. by adding the right number of 'None's for missing 'Optional' fields)
-        Control.Value(record.copy(id = templateId))
+        // Here we extend values of predecessor template types by adding the
+        // right number of 'None's for missing 'Optional' fields
+        val values = new util.ArrayList[SValue]
+        discard(values.addAll(record.values))
+        fields.drop(record.values.size) foreach {
+          case (name @ _, TOptional(_)) => discard(values.add(SOptional(None)))
+          case _ => crash("aaaa")
```

Comment:
aaaaaaaaaaaaaaaaaa

Reply:
Looking at it today, I realized we don't have `Type` in the `compiler/scenario-service/protos/scenario_service.proto`, so it would be some work to add that in the error. Also, since we expect to make this a compile-time and upload-time check, I don't think the runtime error needs to be properly typed (cf. `SBRecUpd` which contains a runtime assertion on the `.id` field)